### PR TITLE
Link earliest version with py.typed in stubsabot obsoletion PRs

### LINKED ISSUE #8754: stubsabot obsoletion PRs: erroneous assumptions that the latest release is the one that added a `py.typed`
See https://github.com/python/typeshed/pull/8746#discussion_r972984030.

`termcolor` releases have been coming thick and fast recently, since maintenance of the package has just been taken over by a new team. There have been two releases in the last week, and stubsabot erroneously assumed that it was the most recent release that added a `py.typed` when making a PR to mark `types-termcolor` as obsolete. (In reality, it was the one before that which added the `py.typed` file).

### PR DESCRIPTION
**Pull request by AlexWaygood**
Fixes #8754.

In order to fix that bug, this PR introduces a new abstraction, a `PypiReleaseDownload` class. Instances of this class represent specific releases of a PyPI distribution, and various pieces of information relevant to that specific release. The `version` and `upload_date` attributes have been moved from the `PypiInfo` class to this new class, since now we potentially need to download multiple versions of the same package from PyPI.

### CODE REVIEW

File: scripts/stubsabot.py [hunk 1]
```diff
@@ -132,21 +151,28 @@ def __str__(self) -> str:
...
         with tarfile.open(fileobj=body, mode="r:gz") as zf:
             return any(Path(f).name == "py.typed" for f in zf.getnames())
     else:
-        raise AssertionError(f"Unknown package type: {packagetype}")
+        raise AssertionError(f"Unknown package type: {packagetype!r}")
+
+
+async def find_first_release_with_py_typed(pypi_info: PypiInfo, *, session: aiohttp.ClientSession) -> PypiReleaseDownload:
+    get_earlier_release: Callable[[], PypiReleaseDownload] = pypi_info.releases_in_descending_order().__next__
+    while await release_contains_py_typed(release := get_earlier_release(), session=session):
```
[the remark this answers is not in the document]
Comment (AlexWaygood) [reply scripts/stubsabot.py:173]: *Grumbles about how people give direct access of dunders too much hate, one of the `itertools` recipes nearly does this exact thing*

File: scripts/stubsabot.py [hunk 2]
```diff
@@ -263,33 +291,42 @@ async def determine_action(stub_path: Path, session: aiohttp.ClientSession) -> U
...
+    latest_release = pypi_info.get_latest_release()
+    latest_version = latest_release.version
     spec = packaging.specifiers.SpecifierSet(f"=={stub_info.version_spec}")
-    if pypi_info.version in spec:
+    if latest_version in spec:
         return NoUpdate(stub_info.distribution, "up to date")
 
+    is_obsolete = await release_contains_py_typed(latest_release, session=session)
+    if is_obsolete:
+        first_release_with_py_typed = await find_first_release_with_py_typed(pypi_info, session=session)
```
[the remark this answers is not in the document]
Comment (AlexWaygood) [reply scripts/stubsabot.py:302]: > this means we'll download six months of versions for all obsolete packages

Whoops

[the remark this answers is not in the document]
Comment (AlexWaygood) [reply scripts/stubsabot.py:303]: > this means we'll download six months of versions for all obsolete packages, even if they're already marked as obsolete.

Wait, does it? Surely if they're already marked as obsolete, we'll already have exited this function due to the check in https://github.com/python/typeshed/pull/8775/files#diff-d11170d18a48b3b03315425963aaba1cf9b7ac91af9488f4e0416a9d9a42bd29R289-R290?

File: scripts/stubsabot.py [hunk 3]
```diff
@@ -71,32 +72,50 @@ def read_typeshed_stub_metadata(stub_path: Path) -> StubInfo:
...
 @dataclass
-class PypiInfo:
-    distribution: str
+class PypiReleaseDownload:
+    url: str
+    packagetype: Annotated[str, "Should hopefully be either 'bdist_wheel' or 'sdist'"]
+    filename: str
     version: packaging.version.Version
     upload_date: datetime.datetime
-    # https://warehouse.pypa.io/api-reference/json.html#get--pypi--project_name--json
```
[the remark this answers is not in the document]
Comment (AlexWaygood) [reply scripts/stubsabot.py:78]: Added it back in 84f9200495b0006f02caf70357e315f5804853a7

**Review by hauntsaninja — APPROVED** [review]
Looks good!

**Pull request merged**